generators/src/bin/wheel-generator.rs

```diff
@@ -273,7 +273,7 @@ pub unsafe fn hardcoded_sieve(bytes: &mut [u8], si_: &mut usize, wi_: &mut usize
                      bit = info.unset_bit, indent = indent);
         }
         println!("
-{indent}    p = p.offset(prime_ * {} + {})
+{indent}    p = (p as usize).saturating_add(prime * {} + {}) as *mut u8;
```

Comment:
Why are we converting to `usize` here? What's wrong with `wrapping_offset`?

Reply:
I'm feeling paranoid that the pointer may actually **wrap**, such that `p < end` is still true while being wildly out of bounds. We could let it wrap and also check `p >= start`, but I found that performed worse than this `saturating_add`.

In practice, wrapping is improbable because the kernel usually owns the entire upper address space. But IIRC, a 32-bit process on a 64-bit kernel is allowed to use almost all of its 4GB 32-bit address space, so it might get close. I suppose `primal` could be used in a kernel or embedded environment too, if you're adventurous. :)

Reply:
Ah, I see.

The potential problem with round-tripping through a `usize` is that currently it confuses Miri (you're actually told that the code is UB), and it's not clear if this code as-written will ever "just work" under Stacked Borrows. The crux of the problem is that it's not clear what the provenance is of the pointer that comes out of that int-to-pointer cast. There are ideas on how to solve this, but that's a bit besides the point. If you use this implementation, you can only run Miri with `-Zmiri-disable-stacked-borrows`. That might be okay for you. I think the issue that prompted this change would still be uncovered with that flag, because the issue is actually reported by the rustc const eval machinery.

Normally here is where I would suggest some alternative but pointers are the darnedest things. So if you want to keep this saturating check I'd just like to see a comment that tries to explain why there's this whole dance going on.

Reply:
Hmm, I ran with `cargo +nightly miri test` and did not see any warnings or errors. Can you share how "confuse" Miri with this?

I could also try manual saturation, something like:

```rust
let p2 = p.wrapping_offset(...);
p = if p < p2 { p2 } else { end };
```

If LLVM realized that this `p < p2` corresponds to the overflow flag, this might have about the same cost as `saturating_add`.

Reply:
You need to set `MIRIFLAGS="-Zmiri-tag-raw-pointers"`. By default it implements an incomplete form of Stacked Borrows, and I never did get around to asking Ralf why.